### PR TITLE
Fix public RFID prefix data leak

### PR DESCRIPTION
**Pull request by arthexis**
### Motivation
- The tokenized public RFID usage endpoint used relaxed prefix matching (`__istartswith`) which caused unrelated cards sharing the same UID prefix to be included in public usage pages, leaking other cardholders' transactions and scan attempts.

### Description
- Replace broad `rfid__istartswith` predicates with exact matches for canonical values and exact truncated reader reports so only the resolved RFID (or an exact truncated reader value) is exposed; the lookup now adds truncated prefixes only as `__iexact` when they represent shorter reader reports. (`apps/cards/public_usage.py`)
- Restrict attempt and transaction query paths to use the tightened predicate so public pages cannot surface other cards' data. (`apps/cards/public_usage.py`)
- Add a regression test ensuring a second RFID with the same 8-character prefix does not contribute transactions, kWh totals, rejection counts, or recent attempts to another card's public token context. (`apps/cards/tests/test_public_card_usage.py`)

### Testing
- Bootstrapped the environment with `./install.sh` and dependency refresh via `./env-refresh.sh --deps-only` to create the virtualenv and install test dependencies, and started Redis as required; environment setup completed successfully.
- Ran the card usage tests with `.venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py` and observed all tests passing: `12 passed`.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a2d731988d08326b83d4168ec48c116)

Refs #8587

## Summary

Fixes a public RFID prefix data leak where the public usage endpoint could expose transactions, kWh totals, rejection counts, and recent attempts belonging to other cardholders that share an RFID UID prefix. The public pages now only attribute data to the resolved RFID (or an exact, unambiguous truncated reader value) by excluding any RFID values/prefixes that are claimed or represent resolver-prefix overlaps for other cards.

## Changes

**apps/cards/public_usage.py**
- Replaced broad prefix matching with tag-aware, exact/unambiguous candidate query construction.
- Added logic to determine whether normalized RFID values/prefix-derived candidates are owned by other cards (including handling reversed UID/prefix overlaps) and to detect ambiguous resolver-prefix overlaps.
- Rebuilt `_rfid_value_query` to be tag-aware (`tag` parameter):
  - it builds ORM filters only from safe, unclaimed exact candidates and prefix-derived candidates,
  - if no safe candidates remain, it returns an empty result filter.
- Updated attempt and transaction query paths to apply the tightened predicate by passing the current `tag` into `_rfid_value_query`.

**apps/cards/tests/test_public_card_usage.py**
- Extended the public-card RFID usage test suite with focused regression/edge-case coverage for:
  - same-8-character prefix collisions (ensuring data from the second RFID does not appear in another card’s public token context),
  - OCPP idTag prefix-length behavior (`OCPP_ID_TAG_LENGTH`),
  - reversed-only overlap handling and reversed/canonical reversed exclusions,
  - shared prefix exclusion and “prefix ownership” boundaries (shorter/mid-length cards owning prefixes that should exclude longer-card matches),
  - ambiguous resolver-prefix scenarios (exact and OCPP-derived),
  - verification that excluded scenarios produce zero totals and do not populate `recent_transactions` / `recent_attempts`.

**tests/test_role_app_profiles.py**
- Improved test isolation by explicitly clearing relevant environment-variable keys in `test_control_role_profile_migration_graph_loads_tmp_path`.

## Testing

- Bootstrapped environment (`./install.sh`, `./env-refresh.sh --deps-only`) and started Redis.
- Ran: `.venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py` — all 12 tests passed.

### CODE REVIEW

File: apps/cards/public_usage.py [hunk 1]
```diff
@@ -22,14 +22,13 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
     prefixes = {
-        candidate[:prefix_length]
+        candidate[: RFID.MATCH_PREFIX_LENGTH]
         for candidate in candidates
-        if candidate[:prefix_length]
+        if len(candidate) > RFID.MATCH_PREFIX_LENGTH
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        query |= Q(**{f"{field_name}__iexact": prefix})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:149]: PatchGate addressed this by making public RFID prefix matching skip any short reader prefix that is already claimed by a different RFID card. Added a regression where a long UID shares its 8-character prefix with a real short-card UID, and the long card's public usage context now excludes the short card's transaction, rejected attempt, totals, and recent rows. Validation: .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (13 passed); py_compile; ruff; git diff --check.

File: apps/cards/public_usage.py [hunk 2]
```diff
@@ -22,21 +26,22 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
+        candidate[: RFID.MATCH_PREFIX_LENGTH]
         for candidate in candidates
-        if candidate[:prefix_length]
+        if len(candidate) > RFID.MATCH_PREFIX_LENGTH
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        if _claimed_by_other_card(prefix, tag):
+            continue
+        query |= Q(**{f"{field_name}__iexact": prefix})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: PatchGate tightened this further so a prefix-only reader report is considered unsafe when any other registered RFID starts with that prefix, including another long UID. Added a regression with two long RFIDs sharing the same 8-character prefix and a prefix-only transaction/attempt labelled to the other card; the public context now excludes those rows and totals. Validation: .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (14 passed); py_compile; ruff; git diff --check.

File: apps/cards/public_usage.py [hunk 3]
```diff
@@ -11,7 +11,11 @@
 from apps.cards.models import RFID, RFIDAttempt
 
 
-def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
+def _prefix_claimed_by_other_card(prefix: str, tag: RFID) -> bool:
+    return RFID.objects.filter(rfid__istartswith=prefix).exclude(pk=tag.pk).exists()
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: PatchGate addressed this in caa92a864 by normalizing prefix ownership checks and treating reversed UID prefixes claimed by another RFID as ambiguous before trusting a prefix-only reader report. Added a regression where two cards share the same reversed-reader prefix while their stored RFIDs differ; the public usage context now excludes the transaction, rejected attempt, totals, and recent rows. Validation: .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (15 passed); ruff; git diff --check.

File: apps/cards/public_usage.py [hunk 4]
```diff
@@ -22,21 +33,22 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
-        candidate[:prefix_length]
+        candidate[: RFID.MATCH_PREFIX_LENGTH]
         for candidate in candidates
-        if candidate[:prefix_length]
+        if len(candidate) > RFID.MATCH_PREFIX_LENGTH
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        if _prefix_claimed_by_other_card(prefix, tag):
+            continue
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: PatchGate addressed this in ef03eb796 by checking exact reversed candidate values against other RFID records before including them in the public usage query. Added a regression with byte-order-reversed cards where the other card's canonical UID is no longer included in the first card's public totals, rejected attempts, or recent rows. Validation: .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (16 passed after isolated retry); ruff; git diff --check. Note: a concurrent Django test startup on another worktree reproduced the local SQLite setup disk I/O issue, so I reran this suite isolated.

File: apps/cards/public_usage.py [hunk 5]
```diff
@@ -21,22 +43,25 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
         candidates.append(reversed_uid)
     query = Q()
     for candidate in candidates:
+        if candidate != normalized and _value_claimed_by_other_card(candidate, tag):
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: PatchGate addressed this in f06811288 by treating every exact RFID candidate, including the canonical value, as unsafe when another registered card claims it directly or as a reversed UID. The query now returns an explicit empty predicate when all candidates are ambiguous instead of accidentally matching all rows. Added a regression for the canonical-value-as-other-reversed-UID case. Validation: .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (17 passed); ruff; git diff --check.

File: apps/cards/public_usage.py [hunk 6]
```diff
@@ -11,7 +11,29 @@
...
+    )
+
+
+def _value_claimed_by_other_card(value: str, tag: RFID) -> bool:
+    normalized = RFID.normalize_code(value)
+    if not normalized:
+        return False
+    return (
+        RFID.objects.exclude(pk=tag.pk)
+        .filter(Q(rfid__iexact=normalized) | Q(reversed_uid__iexact=normalized))
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: PatchGate addressed this in a91f8e828 by making exact RFID candidates ambiguous when another registered raw or reversed UID starts with that value, not only when it equals it. Added a regression where a short public card value is also a longer card truncated reader report, and the short card public page now excludes that transaction and attempt. Validation: ARTHEXIS_SQLITE_PATH=/tmp/... ARTHEXIS_SQLITE_TEST_PATH=/tmp/... .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (18 passed); ruff; git diff --check.

File: apps/cards/public_usage.py [hunk 7]
```diff
@@ -20,23 +42,31 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
+        candidate[: RFID.MATCH_PREFIX_LENGTH]
         for candidate in candidates
-        if candidate[:prefix_length]
+        if len(candidate) > RFID.MATCH_PREFIX_LENGTH
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        if _prefix_claimed_by_other_card(prefix, tag):
+            continue
+        query |= Q(**{f"{field_name}__iexact": prefix})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:149]: Addressed in 9f01a1d6d2e8f5ba5a2976cbbf55d9bf3dd8016a.

Changes:
- Added safe 20-character OCPP idTag prefix matching for long RFID public usage.
- Kept the existing ambiguity guards so shared prefixes claimed by other cards are still skipped.
- Added a regression for a long UID whose transaction and attempt are stored under the 20-character OCPP idTag.

Validation:
- .venv/bin/python -m ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py
- ARTHEXIS_SQLITE_PATH=/tmp/arthexis-pr8729-runtime-rerun.sqlite3 ARTHEXIS_SQLITE_TEST_PATH=/tmp/arthexis-pr8729-test-rerun.sqlite3 .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (20 passed)

[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:149]: Addressed in 9f01a1d6d.

Changes:
- Restored public usage matching for OCPP 20-character idTag prefixes on long released RFIDs while keeping shared 20-character prefixes ambiguous when another card claims them.

Validation:
- ARTHEXIS_SQLITE_PATH=/tmp/... ARTHEXIS_SQLITE_TEST_PATH=/tmp/... .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (20 passed); ruff; py_compile; git diff --check.

File: apps/cards/public_usage.py [hunk 8]
```diff
@@ -10,8 +10,32 @@
...
+    )
+
+
+def _value_claimed_by_other_card(value: str, tag: RFID) -> bool:
+    normalized = RFID.normalize_code(value)
+    if not normalized:
+        return False
+    return (
+        RFID.objects.exclude(pk=tag.pk)
+        .filter(Q(rfid__istartswith=normalized) | Q(reversed_uid__istartswith=normalized))
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Resolved in 4c0a8c680 by treating exact RFID candidates as ambiguous when another registered raw or reversed UID owns one of the matching prefixes, covering the long UID with short-card prefix owner case. Added a regression where the stored transaction and attempt use the long UID but suite matching belongs to the shorter prefix card. Validation: black --target-version py312 --check; ruff check; manage.py test run -- apps/cards/tests/test_public_card_usage.py (21 passed).

File: apps/cards/public_usage.py [hunk 9]
```diff
@@ -10,8 +10,46 @@
...
+
+
+def _value_claimed_by_other_card(value: str, tag: RFID) -> bool:
+    normalized = RFID.normalize_code(value)
+    if not normalized:
+        return False
+    prefixes = {
+        normalized[:prefix_length]
+        for prefix_length in (OCPP_ID_TAG_LENGTH, RFID.MATCH_PREFIX_LENGTH)
+        if len(normalized) > prefix_length
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Resolved in bd66f9d98 by checking whether any other registered raw or reversed UID value with the matching RFID prefix is a shorter owner of the exact candidate, not only fixed 8/20-character prefixes. Added the mid-length prefix regression. Validation: black --target-version py312 --check; ruff check; manage.py test run -- apps/cards/tests/test_public_card_usage.py (22 passed).

File: apps/cards/public_usage.py [hunk 10]
```diff
@@ -20,23 +65,32 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
         for candidate in candidates
-        if candidate[:prefix_length]
+        for prefix_length in (OCPP_ID_TAG_LENGTH, RFID.MATCH_PREFIX_LENGTH)
+        if len(candidate) > prefix_length
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        if _prefix_claimed_by_other_card(prefix, tag):
+            continue
+        query |= Q(**{f"{field_name}__iexact": prefix})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:149]: Addressed in edde80a99. The OCPP 20-character prefix branch now applies the same shorter-owner ambiguity check used for exact candidates, so idTag prefixes owned by shorter raw or reversed RFIDs are skipped. Added a regression where a longer public RFID's OCPP idTag is owned by a mid-length card. Validation: black --target-version py312 --fast --check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ARTHEXIS_SQLITE_PATH=/tmp/... ARTHEXIS_SQLITE_TEST_PATH=/tmp/... .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (23 passed); git diff --check.

File: apps/cards/public_usage.py [hunk 11]
```diff
@@ -10,8 +10,53 @@
...
+    if not prefix_length:
+        return False
+    match_prefix = value[:prefix_length]
+    matches = RFID.objects.exclude(pk=tag.pk).filter(
+        Q(rfid__istartswith=match_prefix) | Q(reversed_uid__istartswith=match_prefix)
+    )
+    for rfid, reversed_uid in matches.values_list("rfid", "reversed_uid"):
+        for owned_value in (rfid, reversed_uid):
+            normalized = RFID.normalize_code(owned_value)
+            if normalized and normalized != value and value.startswith(normalized):
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:49]: Addressed in 24ed4f52a. The public RFID usage matcher now rejects exact and OCPP-derived values whenever another card shares the same raw/reversed resolver prefix, matching the first-8 ambiguity from RFID.matching_queryset(). Added regressions for exact UID and OCPP idTag values with same-prefix collisions. Validation: .venv/bin/ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (25 passed).

File: apps/cards/public_usage.py [hunk 12]
```diff
@@ -10,8 +10,82 @@
...
+
+    query = Q()
+    has_prefix = False
+    for candidate in candidates:
+        prefix_length = min(len(candidate), RFID.MATCH_PREFIX_LENGTH)
+        if not prefix_length:
+            continue
+        resolver_prefix = candidate[:prefix_length]
+        query |= Q(rfid__istartswith=resolver_prefix) | Q(
+            reversed_uid__istartswith=resolver_prefix
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in 8a002aa36. The resolver-prefix ambiguity helper now mirrors RFID.matching_queryset() by checking other rows' rfid prefixes only, so a suffix-only reversed_uid collision no longer hides a card's own full canonical UID usage. Added a regression for the 11111111FACEB00C / 22222222FACEB00C case. Validation: .venv/bin/ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (26 passed).

File: apps/cards/public_usage.py [hunk 13]
```diff
@@ -10,8 +10,80 @@
...
+
+
+def _value_claimed_by_other_card(value: str, tag: RFID) -> bool:
+    normalized = RFID.normalize_code(value)
+    if not normalized:
+        return False
+    directly_claimed = (
+        RFID.objects.exclude(pk=tag.pk)
+        .filter(
+            Q(rfid__istartswith=normalized) | Q(reversed_uid__istartswith=normalized)
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in ad7aac3b. The direct ownership check now treats another card's reversed_uid as a claim only on exact match, not prefix match, so exact short UIDs such as 0CB0CEFA remain visible when they only prefix another card's reversed UID. Added a regression for that case. Validation: .venv/bin/ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; .venv/bin/python manage.py test run -- apps/cards/tests/test_public_card_usage.py (27 passed).

File: apps/cards/public_usage.py [hunk 14]
```diff
@@ -20,23 +90,39 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
         candidate[:prefix_length]
         for candidate in candidates
-        if candidate[:prefix_length]
+        for prefix_length in (OCPP_ID_TAG_LENGTH, RFID.MATCH_PREFIX_LENGTH)
+        if len(candidate) > prefix_length
     }
     for prefix in prefixes:
-        query |= Q(**{f"{field_name}__istartswith": prefix})
+        if (
+            _prefix_claimed_by_other_card(prefix, tag)
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in 6d4f786a13f9.

Changes:
- Allowed 20-character OCPP idTag prefixes when the only overlap is another card's reversed_uid, while preserving 8-character resolver-prefix safeguards.

Validation:
- black --target-version py312 --fast apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; manage.py test run -- apps/cards/tests/test_public_card_usage.py (28 passed)

File: apps/cards/public_usage.py [hunk 15]
```diff
@@ -10,8 +10,80 @@
...
+    if not prefix_length:
+        return False
+    match_prefix = value[:prefix_length]
+    matches = RFID.objects.exclude(pk=tag.pk).filter(
+        Q(rfid__istartswith=match_prefix) | Q(reversed_uid__istartswith=match_prefix)
+    )
+    for rfid, reversed_uid in matches.values_list("rfid", "reversed_uid"):
+        for owned_value in (rfid, reversed_uid):
+            normalized = RFID.normalize_code(owned_value)
+            if normalized and normalized != value and value.startswith(normalized):
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in a90f242945f7.

Changes:
- Stopped treating reversed-only UID prefixes as ownership for exact values and 20-character OCPP idTags while preserving 8-character resolver-prefix collision checks.

Validation:
- black --target-version py312 --fast apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; isolated manage.py test run -- apps/cards/tests/test_public_card_usage.py (30 passed).

File: apps/cards/public_usage.py [hunk 16]
```diff
@@ -20,23 +103,47 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
+                tag,
+                include_reversed_uid=prefix_length == RFID.MATCH_PREFIX_LENGTH,
+            )
+            or _value_shares_resolver_prefix_with_other_card(
+                prefix,
+                tag,
+            )
+        ):
+            continue
+        query |= Q(**{f"{field_name}__iexact": prefix})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:149]: Addressed in 27da7f8250dc.

Changes:
- Matched longer OCPP idTags for cards stored under short RFIDs by adding guarded startswith matching for short canonical candidates, with a regression for FACEB00C plus a longer stored transaction idTag.

Validation:
- black --target-version py312 --fast apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; manage.py test run -- apps/cards/tests/test_public_card_usage.py (31 passed)

File: apps/cards/public_usage.py [hunk 17]
```diff
@@ -20,23 +103,49 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
         candidates.append(reversed_uid)
     query = Q()
+    has_candidate = False
     for candidate in candidates:
+        if _value_claimed_by_other_card(candidate, tag):
+            continue
         query |= Q(**{f"{field_name}__iexact": candidate})
-    prefix_length = min(len(normalized), RFID.MATCH_PREFIX_LENGTH)
+        if len(candidate) <= RFID.MATCH_PREFIX_LENGTH:
+            query |= Q(**{f"{field_name}__istartswith": candidate})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in 09b4eb352e3b.

Changes:
- Matched longer OCPP idTags for stored RFIDs up to the OCPP idTag length after the existing ownership guard accepts the candidate, with mid-length RFID coverage.

Validation:
- black --target-version py312 --fast apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; isolated manage.py test run -- apps/cards/tests/test_public_card_usage.py (32 passed).

[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: ﻿Addressed in 09b4eb352e3b.

Changes:
- Matched longer OCPP idTags for mid-length stored RFIDs by allowing guarded public usage prefix matching up to the OCPP idTag length

Validation:
- black --target-version py312 --check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; ARTHEXIS_SQLITE_TEST_PATH=... manage.py test run -- apps/cards/tests/test_public_card_usage.py (32 passed)

File: apps/cards/public_usage.py [hunk 18]
```diff
@@ -20,23 +103,49 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
         candidates.append(reversed_uid)
     query = Q()
+    has_candidate = False
     for candidate in candidates:
+        if _value_claimed_by_other_card(candidate, tag):
+            continue
         query |= Q(**{f"{field_name}__iexact": candidate})
-    prefix_length = min(len(normalized), RFID.MATCH_PREFIX_LENGTH)
+        if len(candidate) <= OCPP_ID_TAG_LENGTH:
+            query |= Q(**{f"{field_name}__istartswith": candidate})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py:112]: ﻿Addressed in 5a511b2f95f7.

Changes:
- Preserved safe same-resolver-prefix OCPP idTags for mid-length stored RFIDs after the existing ambiguity guard confirms no other card owns that resolver prefix

Validation:
- black --target-version py312 --check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; ARTHEXIS_SQLITE_TEST_PATH=... manage.py test run -- apps/cards/tests/test_public_card_usage.py (33 passed)

File: apps/cards/public_usage.py [hunk 19]
```diff
@@ -20,23 +103,57 @@ def _rfid_value_query(value: str, *, field_name: str = "rfid") -> Q:
...
+            if len(candidate) > RFID.MATCH_PREFIX_LENGTH and candidate == normalized:
+                query |= Q(
+                    **{
+                        f"{field_name}__istartswith": candidate[
+                            : RFID.MATCH_PREFIX_LENGTH
+                        ]
```
[the remark this answers is not in the document]
Comment (arthexis) [reply apps/cards/public_usage.py]: Addressed in 1751ca82bfbe.

Changes:
- Guarded the added 8-character resolver-prefix public match with the same ownership checks used for generated prefixes, so reversed-UID owners block prefix-only rows before they reach the public usage query.

Validation:
- black --target-version py312 --fast apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; ruff check apps/cards/public_usage.py apps/cards/tests/test_public_card_usage.py; git diff --check; isolated manage.py test run -- apps/cards/tests/test_public_card_usage.py (34 passed).
